pytorch/pytorch
# Composite compliance for index_copy, index_fill, masked_scatter, masked_fill

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71766
* #71765
* __->__ #71751
* #71750

Before this PR, each of the above operations were composite and had
in-place variants that were primitive w.r.t. autograd.

The backward passes are not composite compliant due to the op (e.g.
index_copy) decomposing into index_copy_ and index_copy_'s backward
formula having in-place operations in it. To fix this, for each of the
ops mentioned in the title:
- I deleted the autograd formula for the in-place variant and replaced
it with the out-of-place variant
- This makes the forward-ad formula slightly slower because the codegen
generates a temporary but offline discussion suggests it's not worth
maintaining two sets of formulas for this and we can make the autograd
codegen smarter in the future.
- I then replaced instances of grad.clone().inplace_variant_(...) with
grad.outplace_variant(...)

Test Plan:
- run existing tests to check correctness
- run functorch tests

Differential Revision: [D33772756](https://our.internmc.facebook.com/intern/diff/D33772756)
